packages/tools-iottwinmaker/src/lib/workspace.ts

```diff
@@ -0,0 +1,182 @@
+    PolicyDocument: rolePolicy,
+    Description: `${roleType} Policy generated by AWS IoT TwinMaker Getting Started Script`,
+  });
+
+  const policyArn: string | undefined = createPolicyResponse.Policy?.Arn;
+  console.log(`Created Policy: ${policyArn}`);
+  await aws().iam.attachRolePolicy({
+    RoleName: roleName,
+    PolicyArn: policyArn,
+  });
+  console.log(`Attached Policy: ${policyArn} to Role: ${roleArn}.`);
+  return roleArn;
+}
+
+/**
+ * Helper function during workspace creation to create S3 bucket
+ * @param uniqueId identifier for bucket used as suffix in the name
+ * @returns promise of the role S3 bucket name and arn
+ */
+async function createWorkspaceS3Bucket(uniqueId: string) {
+  const s3BucketName = `iot-twinmaker-workspace-bucket-${uniqueId}`;
+  console.log('Creating S3 Bucket for the TwinMaker Workspace...');
+  const createBucketResponse: CreateBucketCommandOutput = await aws().s3.createBucket({ Bucket: s3BucketName });
+  console.log(`Workspace S3 Bucket created: ${createBucketResponse.Location}`);
+  console.log('Configuring CORS policy...');
+  aws().s3.putBucketCors({
+    Bucket: s3BucketName,
+    CORSConfiguration: {
+      CORSRules: [
+        {
+          AllowedHeaders: ['*'],
+          AllowedMethods: ['GET', 'PUT', 'POST', 'DELETE', 'HEAD'],
+          AllowedOrigins: ['*'],
+          ExposeHeaders: ['ETag', 'Access-Control-Allow-Origin'],
+        },
+      ],
+    },
+  });
+  return { s3BucketName, s3BucketArn: `arn:aws:s3:::${s3BucketName}` };
+}
+
+/**
+ * Driver function for workspace creation
+ * @param workspaceId workspace-id to be created
+ * @returns promise of workspace arn, s3 arn, role arn, dashboardrole arn
+ */
+async function prepareWorkspace(workspaceId: string) {
+  const uniqueId: string = createUniqueId();
+  const region: string = aws().region;
+  const { accountId, accountArn } = await aws().getCurrentIdentity();
+  const { s3BucketArn } = await createWorkspaceS3Bucket(uniqueId);
+
+  const policyParams: Record<string, string> = {
+    accountId,
+    accountArn,
+    region,
+    workspaceS3BucketArn: s3BucketArn,
+  };
+
+  const workspaceRolePolicy: string = replaceTemplateVars(JSON.stringify(workspaceRolePolicyTemplate), policyParams);
+  const workspaceRoleArn: string = await createRoleAndPolicy(
+    uniqueId,
+    'WorkspaceRole',
+    JSON.stringify(WORKSPACE_ROLE_ASSUME_POLICY),
+    workspaceRolePolicy
+  );
+
+  const workspaceArn: string = await retryFunction(async () => {
+    const createWorkspaceResponse: CreateWorkspaceCommandOutput = await aws().tm.createWorkspace({
+      workspaceId,
+      role: workspaceRoleArn,
+      s3Location: s3BucketArn,
+    });
+    const workspaceArn: string | undefined = createWorkspaceResponse.arn;
+    if (!workspaceArn) {
+      throw new Error('Failed to create Workspace');
+    }
+    console.log(`Created Workspace: ${workspaceArn}`);
+    return workspaceArn;
+  }, 10);
```

Comment:
later versions of the AWS SDK may support this feature of "waiters" (intended to make it easier to wait for a resource to transition to some state [[1]](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/IoTTwinMaker/Waiters.html) [[2]](https://aws.amazon.com/blogs/developer/waiters-in-modular-aws-sdk-for-javascript/) not sure if they already exist in the client we're using but may be something to check out if it saves us from having to build our own retry logic

Reply:
Interesting, this may be something to look into more but I simplified the retry function as per Matt's comments above, and I think it is fairly straightforward for our use case for now. Not entirely sure if we are using V2 or V3 like linked above